server/team_matchmaker/player_party.py

```diff
@@ -0,0 +1,43 @@
+from server.players import Player
+
+
+class PlayerParty:
+    def __init__(self, team_matchmaking_service: "TeamMatchmakingService", owner: Player):
```

Comment:
Lets try our best to avoid the cyclic dependency of `TeamMatchmakingService <-> PlayerParty` if possible. Matchmmaking service can add parties to it's own member variables, and I think we should figure out a different way of passing messages to the player so that PlayerParty doesn't need to know about the lobbyconnection or protocol at all. Basically try to decouple this a bit more so it can function even if we remove lobbyconnection

Reply:
teammatchmakingservice is a service, PlayerParty is an object. I don't see any issue with the objects managed by a service having a reference to call back that service. In this case it makes it extremly easy to trigger certain actions on adding/remove players. Also the team matchmaking service doesn't have to send data each time it changes something about the groups, the group updates are sent automatically.

Reply:
Well here are specifically the things that bother me:
1. This function is directly modifying a member variable of `team_matchmaking_service` (feature envy again). Seems like it would be better for `team_matchmaking_service` to have a function called `add_player_to_party` which does this modification and then calls `party.add_player`.
```python
self.team_matchmaking_service.player_parties[player] = self
```

2. This object is directly coupled to the implementation of `LobbyConnection` by this line. Obviously this is convenient for quickly sending a message to the player, but it's bad design. Stuff like this is the main cause of the mess in the rest of the code, so I'd like to not contribute further to that mess. I'm not sure what the best way of handling this is, but maybe a callback function would be ok? A callback still doesn't feel elegant to me, but it feels a little better than directly coupling to lobbyconnection code. I will keep thinking about possible solutions to this.
```python
player.lobby_connection.send({
    "command": "update_party",
    "owner": self.owner,
     "members": {m.id for m in self._members}
})
```

Reply:
Maybe instead of accessing `player.lobby_connection` directly to send messages, we could add a `send_message` method for `Player`. That way you can still send messages directly to a player if you have a reference to the player object, but you don't need to know about the implementation of lobby_connection.